bite-size-notes/week-1-e-linearization.md

```diff
@@ -0,0 +1,250 @@
+
+$$
+∇_x f(x_0) = \left[ \begin{array}{cc}
+\frac{\partial \dot{x}_1}{\partial x_1} & \frac{\partial \dot{x}_1}{\partial x_2} \\
+\frac{\partial \dot{x}_2}{\partial x_1} & \frac{\partial \dot{x}_2}{\partial x_2}
+\end{array} \right]
+ = \left[ \begin{array}{cc}
+0 & 1 \\
+-\omega^2 \cos(x_1) & -r
+\end{array} \right]
+ = \left[ \begin{array}{cc}
+0 & 1 \\
+-\omega^2 & -r
+\end{array} \right]
+$$
+
+Thus, the linear approximation of $f(x)$ near $x_0 = [0, 0]^T$ would be:
+
+$$
+f(x) ≈ ∇_x f(x_0)^Tx - ∇_x f(x_0)^T x_0 + f(x_0)
+$$
+
+$$
+f(x) ≈ \textcolor{green}{\left[ \begin{array}{cc}
+0 & 1 \\
+-\omega^2 & -r
+\end{array} \right]} * 
+\left[ \begin{array}{c}
+x_1 \\
+x_2
+\end{array} \right] - 
+\textcolor{red}{\left[ \begin{array}{cc}
+0 & 1 \\
+-\omega^2 & -r
+\end{array} \right] *
+\left[ \begin{array}{c}
+0\\
+0
+\end{array} \right] +
+\left[ \begin{array}{c}
+0\\
+0
+\end{array} \right]}
+$$
+
+$$
+f(x) ≈ \textcolor{green}{A} x + \textcolor{red}{C}
+$$
+
+ ![alt text](figs/Pendulum.PNG "Title")
+ 
+- Now, we will use **codes** for the Pendulum example especially **JAX**
+  
+**JAX:** can automatically compute gradients of functions using forward-mode differentiation **('jax.jacfwd)** and reverse-mode differentiation **(jax.grad)**
+The code is as follows:
+```python
+# JAX
+import jax.numpy as jnp
+from jax import grad, jacfwd
+
+def pendulum_dynamics(x, omega=1.0, r=0.1):
+    x1, x2 = x
+    dx1dt = x2
+    dx2dt = -omega**2 * jnp.sin(x1) - r * x2
+    return jnp.array([dx1dt, dx2dt])
+
+def linearize_system(x, omega=1.0, r=0.1): 
+    jacobian_func = jacfwd(pendulum_dynamics, argnums=0)
+    jacobian_at_x = jacobian_func(x, omega, r)   
+    return jacobian_at_x
+
+x_example = jnp.array([0.0, 0.0])  # Equilibrium point
+omega = 1.0  # Natural frequency
+r = 0.1      # Damping coefficient
+jacobian_at_equilibrium = linearize_system(x_example, omega, r)
+print("Jacobian matrix at the equilibrium point:", jacobian_at_equilibrium)
+```
+- Other **Snippet codes** you can use for **linearization**:
+```  
+# TensorFlow
+import tensorflow as tf
+@tf.function
+def pendulum_tf(x):
+    x1, x2 = x[0], x[1]
+    return tf.stack([x2, -omega**2 * tf.sin(x1) - r * x2])
+x_tf = tf.Variable([0.0, 0.0])
+with tf.GradientTape() as tape:
+    tape.watch(x_tf)
+    y_tf = pendulum_tf(x_tf)
+A_tf = tape.jacobian(y_tf, x_tf)
+
+# PyTorch
+import torch
+def pendulum_torch(x):
+    x1, x2 = x.unbind()
+    return torch.stack([x2, -omega**2 * torch.sin(x1) - r * x2])
+x_torch = torch.tensor([0.0, 0.0], requires_grad=True)
+y_torch = pendulum_torch(x_torch)
+A_torch = torch.autograd.functional.jacobian(pendulum_torch, x_torch)
+
+# Auto_diff (Pseudocode)
+def pendulum_auto_diff(x, omega=1.0, r=0.1):
+    x1, x2 = x
+    return [x2, -omega**2 * auto_diff.sin(x1) - r*x2]
+x0_auto_diff = [0.0, 0.0]
+A_auto_diff = auto_diff.compute_jacobian(pendulum_auto_diff, x0_auto_diff)
+```
```

Comment:
Thank you for sharing your idea. In my opinion, since it is bite notes and need to be simple and short, I think just put some snippets and explanation is enough.